### PR TITLE
Feat: telemetry and event headers

### PR DESCRIPTION
**Pull request by asalkeld**

### CODE REVIEW

File: pkg/plugins/events/sns/sns.go [hunk 1]
```diff
@@ -39,16 +41,16 @@ type SnsEventService struct {
 	provider  core.AwsProvider
 }
 
-func (s *SnsEventService) getTopics() (map[string]string, error) {
+func (s *SnsEventService) getTopics(ctx context.Context) (map[string]string, error) {
 	return s.provider.GetResources(core.AwsResource_Topic)
 }
 
-func (s *SnsEventService) getStateMachines() (map[string]string, error) {
+func (s *SnsEventService) getStateMachines(ctx context.Context) (map[string]string, error) {
```
Comment (tjholm) [inline pkg/plugins/events/sns/sns.go:48]: ctx unused?

File: pkg/plugins/events/sns/sns.go [hunk 2]
```diff
@@ -39,16 +41,16 @@ type SnsEventService struct {
 	provider  core.AwsProvider
 }
 
-func (s *SnsEventService) getTopics() (map[string]string, error) {
+func (s *SnsEventService) getTopics(ctx context.Context) (map[string]string, error) {
```
Comment (tjholm) [inline pkg/plugins/events/sns/sns.go:44]: ctx unused?

File: pkg/plugins/gateway/cloudrun/http.go [hunk 3]
```diff
@@ -47,6 +49,9 @@ func middleware(ctx *fasthttp.RequestCtx, pool worker.WorkerPool) bool {
 	if err := json.Unmarshal(bodyBytes, &pubsubEvent); err == nil && pubsubEvent.Subscription != "" {
 		// We have an event from pubsub here...
 
+		topic := pubsubEvent.Message.Attributes["x-nitric-topic"]
+		ctx := span.FromHeaders(context.TODO(), "topic-"+topic, triggers.HttpHeaders(&rc.Request.Header))
```
Comment (tjholm) [inline pkg/plugins/gateway/cloudrun/http.go:53]: Does this work? I can't see where otel propagation would be occurring using HTTP headers for pubsub message publishing. What tells pubsub to apply these HTTP headers to the subscription?

File: pkg/providers/gcp/membrane.go [hunk 4]
```diff
@@ -29,14 +42,57 @@ import (
...
+	if err != nil {
+		return nil, errors.WithMessagef(err, "NITRIC_TRACE_SAMPLE_PERCENT should be an int not %s", rate)
+	}
+
+	return sdktrace.NewTracerProvider(
+		sdktrace.WithSampler(sdktrace.ParentBased(sdktrace.TraceIDRatioBased(rate))),
+		sdktrace.WithResource(res),
+		sdktrace.WithBatcher(exp),
+	), nil
+}
```
Comment (tjholm) [inline pkg/providers/gcp/membrane.go:89]: would be good to do this consistently, probably in a separate file like you have done for aws.

**Review by tjholm — CHANGES_REQUESTED** [review]
Some initial thoughts at first glance. Will need to take a few more passes over.

File: pkg/plugins/events/sns/sns.go (hunk 1 above)
Comment (asalkeld) [reply pkg/plugins/events/sns/sns.go:48]: i'll pass this down into the resourcesGet

**Review by tjholm — APPROVED** [review]

File: contracts/proto/faas/v1/faas.proto [hunk 5]
```diff
@@ -119,6 +122,9 @@ message TriggerRequest {
   // Or rely on context?
   string mime_type = 2;
 
+  // TraceInformation from the membrane
+  TraceContext trace_context = 10;
```
Comment (tjholm) [inline contracts/proto/faas/v1/faas.proto:126]:
```suggestion
  // TraceInformation from the membrane
  // Allows tying traces from external triggers (e.g. HttpRequests)
  // into each event request/response pair of the Bidirectional stream.
  // which cannot be facilitated by OOTB stream interceptors from OTEL.
  TraceContext trace_context = 10;
```

**Review by tjholm — APPROVED** [review]

**Review by HomelessDinosaur — APPROVED** [review]

**Pull request merged**